Add FileHeaderRule

marcelofabri opened the pull request:
Fixes #844

jpsim commented:
Is there a way we can make this rule a bit more generic? Maybe by making it configurable by things that must or must not be in the top comment?

For example, as much as some people might consider the top comment to be superfluous, some projects _require_ an identical header (e.g. [Realm](https://github.com/realm/realm-cocoa/blob/master/RealmSwift/Realm.swift#L1-L17)). And some use a consistent template (SwiftLint).

marcelofabri commented:
Maybe we could add two configurations, like `enforced_pattern` and `forbidden_pattern`?

jpsim commented:
Maybe `required_string`, `required_pattern`, `forbidden_string` and `forbidden_pattern`?

jpsim commented on CHANGELOG.md:
```diff
@@ -29,6 +29,11 @@
   [#910](https://github.com/realm/SwiftLint/issues/910)
   [#911](https://github.com/realm/SwiftLint/issues/911)
+  
+* Add `file_header` opt-in rule that warns when a file contains a 
```
This is now in the wrong changelog section.

jpsim commented on Source/SwiftLintFramework/Models/MasterRuleList.swift:
```diff
@@ -63,6 +63,7 @@ public let masterRuleList = RuleList(rules:
     LegacyConstantRule.self,
     LegacyConstructorRule.self,
     LegacyNSGeometryFunctionsRule.self,
+    FileHeaderRule.self,
```
This isn't alphabetized.

jpsim commented on Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift:
```diff
@@ -0,0 +1,83 @@
+
+    // swiftlint:disable:next force_try
+    private static let defaultRegex = try! NSRegularExpression(pattern: "\\bCopyright\\b",
+                                                              options: [.caseInsensitive])
```
`options` should be vertically aligned with `pattern`.

jpsim commented on Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift:
```diff
@@ -0,0 +1,83 @@
+        ", forbidden_pattern: \(forbiddenPattern)"
+    }
+
+    public init() {
```
`public init() {}`?

jpsim commented on Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift:
```diff
@@ -0,0 +1,83 @@
+                                                    options: [.ignoreMetacharacters])
+        } else if let requiredPattern = configuration["required_pattern"] as? String {
+            self.requiredPattern = requiredPattern
+            requiredRegex = try NSRegularExpression.cached(pattern: requiredPattern)
```
`requiredRegex = try .cached(pattern: requiredPattern)`

jpsim commented on Source/SwiftLintFramework/Rules/RuleConfigurations/FileHeaderConfiguration.swift:
```diff
@@ -0,0 +1,83 @@
+                                                     options: [.ignoreMetacharacters])
+        } else if let forbiddenPattern = configuration["forbidden_pattern"] as? String {
+            self.forbiddenPattern = forbiddenPattern
+            forbiddenRegex = try NSRegularExpression.cached(pattern: forbiddenPattern)
```
`forbiddenRegex = try .cached(pattern: forbiddenPattern)`

marcelofabri commented:
@jpsim I was enabling this rule for the project, but I've found (at least) two patterns:

```
//
//  String+SwiftLint.swift
//  SwiftLint
//
//  Created by JP Simard on 2015-05-16.
//  Copyright (c) 2015 Realm. All rights reserved.
//
```

```
//
//  SyntaxMap+SwiftLint.swift
//  SwiftLint
//
//  Created by Norio Nomura on 2/19/16.
//  Copyright © 2016 Realm. All rights reserved.
//
```

Should me make them consistent? The second one seems the be the Xcode 8.1 default template now. Or should I just make the regex more flexible and allow the first one as well?

jpsim commented:
I'd love if you could address the inconsistencies, actually! Sticking with the newer Xcode 8.1 default seems preferable.

marcelofabri commented:
@jpsim I've addressed the inconsistencies although the date in the template seems to change according to the user's locale. I've changed to what seems the default to the US locale.

jpsim commented on Source/SwiftLintFramework/Rules/FileLengthRule.swift:
```diff
@@ -2,8 +2,8 @@
 //
-//  Created by JP Simard on 2015-05-16.
-//  Copyright (c) 2015 Realm. All rights reserved.
+//  Created by JP Simard on 5/16/15.
```
Some of these dates are now `mm/dd/yy` (like this one), others are now `dd/mm/yy`.

marcelofabri replied:
they should be all in `mm/dd/yy` now

jpsim commented:
👍 feel free to merge once CI agrees.

marcelofabri commented:
@jpsim Do you have any idea on why the tests are passing with Xcode but not SPM? I get the same thing on my machine, so it's not a Travis issue.

jpsim commented:
Looks like it's the `FunctionBodyLengthRuleTests` that are failing due to an unrelated rule (`file_header`). A quick glance seems to confirm my suspicions that these tests don't limit the rules it checks to the `FunctionBodyLengthRule`: https://github.com/realm/SwiftLint/blob/master/Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift

If you pass in a config constructed with [`makeConfig(...)`](https://github.com/realm/SwiftLint/blob/master/Tests/SwiftLintFrameworkTests/TestHelpers.swift#L116-L123) to `violations(...)` it should limit those tests to just the rule being tested.

marcelofabri commented:
@jpsim I've created a config just disabling `file_header` as `makeConfig` was `fileprivate` and I'm not sure if we want to expose it. What do you think?

jpsim commented:
I'd rather just make `makeConfig` `internal`.

marcelofabri commented:
@jpsim do you wanna review the latest commits or can I merge this? 🚢

marcelofabri commented:
🚢🎉

marcelofabri commented on CHANGELOG.md:
```diff
@@ -14,6 +14,11 @@
+* Add `file_header` opt-in rule that warns when a file contains a 
+  copyright comment header, such as the one Xcode adds by default.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#884](https://github.com/realm/SwiftLint/issues/884)
```
@vincentisambart Thanks for letting me know! Fixed in #993
